### PR TITLE
Enumerables: Rebuild methods in Ruby

### PR DESCRIPTION
**Pull request by arthurborgesdev**
Add the following methods:

- #my_each
- #my_each_with_index
- #my_select
- #my_all?
- #my_any?
- #my_none?
- #my_count
- #my_map
- #my_inject

- Modification of #my_map to accept a proc and then either a proc or a block. 

The "Metrics/PerceivedComplexity" and the "Metrics/CyclomaticComplexity" functionalities of Rubocop were disabled because the methods #my_all?, #my_any? and #my_none had to have more complexity due to more ifs and else that were needed to implement edge cases such as Regex and Classes as arguments and blocks passed.

Also, the "Module Length" in `.rubocop.yml` was incremented due to the same reason above: the complex methods made the module larger than 100 lines (102).

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,70 @@
...
+Run `git clone git@github.com:arthurborgesdev/enumerable_methods.git`
+
+
+### Prerequisites
+
+Need Ruby 1.9+ installed
+
+### Usage
+
+`ruby bubble_sort.rb`
```
Comment (cyonii) [inline README.md:35]: Are you sure this instruction is about this project?

**Review by cyonii — CHANGES_REQUESTED** [review]
# Status: Changes Required ♻️✍️

Hello team, you have shown a great effort in this project, most of the tests are passing. 👏

## To Highlight

I'll like to commend for the following:

- Excellent PR description
- Keeping your code clean 💯
- Configuring your linter and having them all green 🟢.
- Your commit messages are great 👏
- For having your code pass most of the test cases in your first review

## To Improve

You have already shown a great effort, but your methods are still not passing all of the test. Kindly look at the review below and implement these changes. You need to pay close attention to the contexts.

- [ ] You seem to have forgotten to include `multiply_els` method.

> **Failed examples:**

**`#my_each_with_index`**
- [x]  when a block is given, when self is an array, it behaves like `#each_with_index` and call the given block once for each element in self
- [x] when a block is given, when self is a range, it behaves like `#each_with_index` and call the given block once for each element in self
- [ ] when a block is given, when self is a hash, it behaves like `#each_with_index` and call the given block once for each element in self

**`#my_all?`**
- [ ] when a `Regex` is passed as an argument, it returns true if all of the collection matches the `Regex`
- [ ] when a pattern other than `Regex` or a `Class` is given, it returns true if all of the collection matches the pattern


**`#my_any?`**
- [ ] when a pattern other than `Regex` or a `Class` is given, it returns true if any of the collection matches the pattern

**`#my_none?`**
- [ ] when a `Regex` is passed as an argument, it returns true if none of the collection matches the `Regex`
- [ ] when a `Regex` is passed as an argument, it returns false if any of the collection matches the `Regex`
- [ ] when a pattern other than `Regex` or a `Class` is given, it returns false only if one of the collection matches the pattern

**`#my_inject`**
- [ ] when a block is given with an initial value, it combines all elements of `enum` by applying a binary operation, specified by a block
- [ ] when a symbol is specified without an initial value, it combines each element of the collection by applying the symbol as a named method
- [ ] when a symbol is specified without an initial value, it combines each element of the collection by applying the symbol as a named method:range
- [ ] when a symbol is specified with an initial value, it combines each element of the collection by applying the symbol as a named method
- [ ] when a symbol is specified with an initial value, it combines each element of the collection by applying the symbol as a named method::range

**`#multiply_els`**
- [x] it accepts an array as an argument and multiplies all the elements of the array together using `#my_inject`.


Happy coding, keep up the energy. 🔋

### You can find me on slack 👉 [CY Kalu](https://microverse-students.slack.com/team/U01494U96FR) if you need further assistance or clarification. I'm always available and will be glad to help.

File: README.md (hunk 1 above)
Comment (arthurborgesdev) [reply README.md:35]: Thank you, we will correct it!

**Review by jubaan — CHANGES_REQUESTED** [review]
[![Microverse][microverse-shield]][microverse-url]
<div>
<h1 align="center"> ⚠️ 🛠️ ⚠️ 🛠️ Changes Required 🛠️ ⚠️ 🛠️ ⚠️ </h1>
<br />
<p><b>"Fail fast, fail often"</b> and don't let this discourage you, because <b>only by failing do we learn to succeed</b>.</p>
<details open="open">
  <summary>Review Summary</summary>
  <ol>
    <li><a href="#congrats"> 💯 Awesome work 👏🏼 👏🏼 </a></li>
    <li><a href="#mandatory"> ⚠️ Things to Improve </a> 🚨 🚨 </li>
    <li><a href="#optional"> 💡Nice to have and other comments ✍🏼 ✍🏼 </a></li>
    <li><a href="#contact"> 🧔🏼 About your Reviewer </a> (you don't know it... but we are just automated bots 🤖) </li>
    <li><a href="#help"> 🆘 <b> 😭 Feeling stuck? or Need help</b>? 🙏🏼 🙏🏼 </a></li>
  </ol>
</details>
</div>

<h3 id="congrats"> 1. :100: Awesome work</h3>
<details close="close">
<summary> [click to open] Big or little achievements, but achievements at the end. <b>Feel proud that you're learning and improving everyday.</b></summary>



</details>

<hr>
<h3 id="mandatory"> 2. ⚠️ Things to Improve (or the To-Do list to get the project approved) </h3> 
<details close="close">
<summary> [click to open] Working on these observations will help you advance and get closer to the goal 🎯, postpone them or do them half-ways and they'll haunt you 👻 👻... FOREVER!! (It will feel like that, trust me). </summary>


- [ ] enumerables #my_all? when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern
- [ ] enumerables #my_any? when a pattern other than Regex or a Class is given returns true if any of the collection matches the pattern
- [ ] enumerables #my_none? when a Regex is passed as an argument returns true if none of the collection matches the Regex
- [ ] enumerables #my_none? when a Regex is passed as an argument returns false if any of the collection matches the Regex
- [ ] enumerables #my_none? when a pattern other than Regex or a Class is given returns false only if one of the collection matches the pattern
- [ ]  enumerables #my_inject when a block is given with an initial value combines all elements of enum by applying a binary operation, specified by a block



</details>
<p><b>NOTE</b>This list is or could be divided in what is show in this comment and inline comments posted below.<p>
<hr>

<h3 id="optional"> 3. 💡 Nice to have and other comments </h3>
<details close="close">
<summary> [click to open] </h3></summary>


- [ ] Nothing more to add.


</details>

<h3 id="contact"> 4. 🧔🏼 About your Reviewer </h3>
<details close="close">
<summary> [click to open] Say "hello world!" on Slack: @jubaan</summary>
<br />
<div align="center">
  <a src="https://www.github.com/jubaan">
    Julio Añoveros
  </a>
  <p>
    Julio Añoveros | @jubaan</br></br>
    <b>Full-stack Web Developer</b> and</br>
    🤖 Code Reviewer at <a href="https://www.microverse.org/?grsf=kfm2ej">Microverse</a>:rocket:
  </p>
</div>
</details>

<h3 id="help"> 5. 🆘 <b> 😭 Feeling stuck? or Need help</b>? 🙏🏼 🙏🏼</h3>
<details close="close">
  <summary> [click to open] <b>Feel stuck?</b> </summary>
  <br>
  <p> First breath... just do it. Take 15-20 min away from the computer, stretch, meditate, do some exercise, drink water, or do whatever is necessary to distract your mind from the obstacle.</p>
  <p><b>Sometimes you just need to get away to see things with a new perspective.</b></p>
</details>

<details close="close">
  <summary> [click to open] <b>Need help?</b> </summary>
  <br>
  <p> Send me a DM in <a href="https://www.microverse.org/?grsf=kfm2ej">Microverse</a> Students Slack, you'll find me as @jubaan. If I'm not available there are other Code Reviewers you can contact. They'll help as well.</p>
  <p>Remember that <a href="https://www.microverse.org/?grsf=kfm2ej">Microverse</a> has other channels where you can get an answer to your questions:</p>
  <ol>
    <li>The <b>students channel 🧑‍🎓 </b> in slack,</li>
    <li>your <b>teammates 👥 </b>,</li>
    <li>your <b>mentor 👨🏼 👩🏼 </b>, and</li>
    <li><b>Student Success 🏫 </b>.</li>
  </ol>
  <p><b> DON'T hesitate to ask! </b></p>
</details>
<p>If all of the above doesn't help... get a <a href="https://rubberduckdebugging.com/">rubber duck</a> ... it would be great to own one 😄. </p>

<h2 align="center"> Keep up the great work :muscle: and Happy Coding :smiley: :computer:!! </h1>

[microverse-shield]: https://img.shields.io/badge/-microverse-%236F23FF?style=for-the-badge
[microverse-url]: https://www.microverse.org/?grsf=kfm2ej
[bootstrap-shield]: https://img.shields.io/badge/Bootstrap-white?style=for-the-badge&logo=bootstrap
[bootstrap-url]: https://www.microverse.org/?grsf=kfm2ej

**Review by cyonii — APPROVED** [review]
# Status: Approved ✅

Hurray 🎉!! You have completed all the requirements for this project and your methods are passing all of the tests 👏. You have done an impressive job.

You can now merge your PR and move on.

Keep up the energy 🔋, and happy coding.

> #### <br>You can ping me on Slack @[CY Kalu](https://microverse-students.slack.com/team/U01494U96FR) if you need to draw my attention to this review. I'm always available and will be glad to help.<br><br>

**Pull request merged**